### PR TITLE
irgen: basic code generation for algebraic data types

### PR DESCRIPTION
**Pull request by eisenstatdavid**
Provides an API similar to the one specified in #16240.

### CODE REVIEW

File: pkg/sql/ir/irgen/codegen/codegen.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+	CapRefs  int
+	LenAlloc int
+}
+
+// GenerateCode writes code implementing the given types into b.
+func GenerateCode(b *bytes.Buffer, s Settings, namedTypes []ir.NamedType) string {
+	if s.LenAlloc <= 0 {
+		panic("LenAlloc must be positive")
+	}
+	fmt.Fprintf(b, "package d\n")
```
Comment (knz) [inline pkg/sql/ir/irgen/codegen/codegen.go]: please consider https://github.com/knz/compose-text

File: pkg/sql/ir/irgen/test/main.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		b2 := ref2.MustBeBinExpr()
+		return b1.Op() == b2.Op() &&
+			DeepEqual(b1.Left(), b2.Left()) && DeepEqual(b1.Right(), b2.Right())
+	default:
+		panic("unknown Expr tag")
+	}
+}
+
+func main() {
+	var a Allocator
```
Comment (knz) [inline pkg/sql/ir/irgen/test/main.go]: turn this into a Go test

**Review by knz — COMMENTED** [review]
I really like the look & feel of the generated code!
seeing this result makes me more and more confident this is a good investment.

My concern right now is similar as during my first attempt at the end of last  year: the code for the printing is very verbose and poorly maintainable. I wonder if we can do better, I made a suggestion to this effect.

Also TC is complaining about a few things.

**Comment by eisenstatdavid** [comment]
Code generation was completely rewritten. WIP because there is still significant documentation effort left: better comments, updating the RFC, and an explanation of the concepts behind `irgen/template`, but @knz this is ready for another look.

**Comment by eisenstatdavid** [comment]
Added documentation and tidied the lint output. This is ready for formal review now, @knz.

**Comment by knz** [comment]
This is very good. Thank you!
:lgtm_strong: 

---

Reviewed 2 of 7 files at r2, 21 of 21 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16857#-:-KpVTMX_376tTzJq5Bfy:b-kx5jj4)*

**Pull request merged**